pkg/helm/watcher/cache/cache.go

```diff
@@ -0,0 +1,193 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/gofrs/flock"
+
+	pb "github.com/weaveworks/weave-gitops/pkg/api/profiles"
+)
+
+const (
+	// we globally lock the whole base cache folder to avoid the
+	// possibility of checking if a sub-folder exists, just so it is created the moment
+	// we try to create it after we checked that it exists.
+	lockFilename    = "cache.lock"
+	profileFilename = "profiles.yaml"
+	valuesFilename  = "values.yaml"
+	lockTimeout     = 1 * time.Minute
+)
+
+// type alias for easier reading of the cache layout for values.
+type profileName = string
+type profileVersion = string
+
+// ValueMap contains easy access for a profile name and version based values file.
+type ValueMap map[profileName]map[profileVersion][]byte
+
+// Cache defines an interface to work with the profile data cacher.
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . Cache
+type Cache interface {
+	Put(ctx context.Context, helmRepoNamespace, helmRepoName string, value Data) error
+	Delete(ctx context.Context, helmRepoNamespace, helmRepoName string) error
+	// ListProfiles specifically retrieve profiles data only to avoid traversing the values structure for no reason.
+	ListProfiles(ctx context.Context, helmRepoNamespace, helmRepoName string) ([]*pb.Profile, error)
+	// GetProfileValues will try and find a specific values file for the given profileName and profileVersion. Returns an
+	// error if said version is not found.
+	GetProfileValues(ctx context.Context, helmRepoNamespace, helmRepoName, profileName, profileVersion string) ([]byte, error)
+}
+
+// Data is explicit data for a specific profile including values.
+// Saved as `profiles.yaml` and `profileName/version/values.yaml`.
+type Data struct {
+	Profiles []*pb.Profile `yaml:"profiles"`
+	Values   ValueMap
+}
+
+// ProfileCache is used to cache profiles data from scanner helm repositories.
+type ProfileCache struct {
+	cacheLocation string
+}
+
+var _ Cache = &ProfileCache{}
+
+// NewCache initialises the cache and returns it.
+func NewCache(cacheLocation string) (*ProfileCache, error) {
+	if err := os.MkdirAll(cacheLocation, 0700); err != nil {
+		return nil, fmt.Errorf("failed to create helm cache dir: %w", err)
+	}
+
+	return &ProfileCache{
+		cacheLocation: cacheLocation,
+	}, nil
+}
+
+// Put adds a new entry or updates an existing entry in the cache for the helmRepository.
+func (c *ProfileCache) Put(ctx context.Context, helmRepoNamespace, helmRepoName string, value Data) error {
+	op := func() error {
```

Comment:
"op" :thinking: can we be more descriptive?

Reply:
short for operation :) but we can, sure

Reply:
operation to do what though :D

Reply:
Well, anything. :D It's a blanket term for a function that is going to be executed with a lock so I don't have to repeat the locking logic everywhere. :D so I don't know. It's a generic operation. Maybeeeee ummm.... lockOperation or cacheOperation or cacheFunc?